Return the AcceptBlock CValidationState directly in ProcessNewBlock

TheBlueMatt opened the pull request:
This is a first step towards a series of cleanups leading to a (more-complete) #16175. It's just two rather-nice (IMO) cleanups.

In practice this means that CheckBlock+ContextualCheckBlock are
called with a passed-in CValidationState before we move onto
connecting the best chain. This makes conceptual sense as these
calls represent the DoS checks on a block (ie PoW and malleability)
which the caller almost certainly wants to know about right away
and shouldn't have to wait on a callback for (and other
validationinterface clients shouldn't care about someone submitting
bogus malleated blocks to PNB).

This also makes it much, much easier to move the best chain
activation logic to a background thread as it implies that if PNB
returns with a IsValid() CValidationState we don't need to care
about trying to process (non-malleated) copies of the block from
other peers.

TheBlueMatt committed:
Remove unnecessary cs_mains in denialofservice_tests

9fdf05d70cac4a62d1aeeb4299e2c3a9a866f8af resolved some lock
inversion warnings in denialofservice_tests, but left in a number
of cs_main locks that are unnecessary (introducing lock inversion
warnings in future changes).

promag commented on src/net_processing.cpp:
```diff
@@ -2902,10 +2910,14 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
-            ProcessNewBlock(chainparams, pblock, /*fForceProcessing=*/true, &fNewBlock);
-            if (fNewBlock) {
+            CValidationState dos_state;
+            ProcessNewBlock(chainparams, pblock, dos_state, /*fForceProcessing=*/true, &fNewBlock);
```
Does it make sense to have a `ProcessAndNotifyNewBlock` to wrap this repeated code?

TheBlueMatt replied:
I'd prefer to just wait and clean it up in a later PR, given I want to move towards doing this in the background anyway.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16279#discussion_r297034762

In commit "Return the AcceptBlock CValidationState directly in ProcessNewBlock" (dd07ec726c5ade950435876d39741d52a33471ce)

> I'd prefer to just wait and clean it up in a later PR, given I want to move towards doing this in the background anyway

Seems to me deduplicating this code now would only simplify the later PR.

TheBlueMatt replied:
Right, but in the full branch I first clean up some cruft, and then merge the now-identical code instead of the currently-partially-identical code. Easier to just leave it as-is.

promag reviewed (commented):
Concept ACK.

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -2813,15 +2817,19 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+                    BlockChecked(*pblock, dos_state, connman);
+                }
                 LOCK(cs_main);
                 mapBlockSource.erase(pblock->GetHash());
```
Why is mapBlockSource.erase call needed here? In not valid case, this should already happen in explicit BlockChecked call. In valid case, I assume it happens in BlockChecked validation callback?

This logic is messy and repetitive and seems like it could be simplified by moving more of it out of here into the BlockChecked() function, or breaking BlockChecked up into reusable parts.

TheBlueMatt replied:
This PR removes the (now redundant) BlockChecked call in case a block is completely bogus (ie !dos_state.IsValid()), so, no.

jnewbery replied:
I think you're right that we can't get rid of this `mapBlockSource.erase()` call, but not for the reason you say. We still need to call it if the block is valid, but this isn't the first time we've received it (ie `fNewBlock` is returned as true). I think this logic would be easier to follow if structured as follows, with comments:

```c++
        ProcessNewBlock(chainparams, pblock, dos_state, forceProcessing, &fNewBlock);
        if (!dos_state.IsValid()) {
            // The block failed anti-dos / mutation checks. Call BlockChecked() callback here.
            // This clears the block from mapBlockSource.
            BlockChecked(*pblock, dos_state, connman);
        } else if (!fNewBlock) {
            // Block was valid but we've seen it before. Clear it from mapBlockSource.
            LOCK(cs_main);
            mapBlockSource.erase(pblock->GetHash());
        } else {
            // Block is valid and we haven't seen it before. set nLastBlockTime for this peer.
            pfrom->nLastBlockTime = GetTime();
        }
```

TheBlueMatt replied:
Hmm, yea, thats a bit cleaner, however, after #16323, which this was refactored out of, the original goes back to being cleaner, so would prefer to just leave it (also cause its fewer LoC changed).

ryanofsky reviewed (commented):
Concept ACK. Changes look good and I should add a review ACK soon, but I need to dig in a little more to understand this better.

TheBlueMatt commented on src/validation.h:
```diff
@@ -198,22 +198,30 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
  *
- * Note that we guarantee that either the proof-of-work is valid on pblock, or
- * (and possibly also) BlockChecked will have been called.
+ * If pblock connects, and has been malleated, state is guaranteed to be some
```
Maybe the confusion was the use of the old term "malleated" instead of the "mutated" term used in ValidationInvalidReason. I changed references to mutated.

TheBlueMatt commented on src/validation.h:
```diff
@@ -198,22 +198,30 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
  *
- * May not be called in a
- * validationinterface callback.
+ * If fForceProcessing is set, barring pruning and a desire to re-download a
```
This is saying something very specific - if fForceProcessing is set (but deliberately not saying anything about *when* you should set it, as that is a rather complicated topic), unless your goal is to re-download a block, you *never* need to call ProcessNewBlock again with the block. That is to say, "this copy of the block is good, no need to try to find an alternate source".

instagibbs commented on src/validation.cpp:
```diff
@@ -3428,14 +3428,13 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
 }
 
-bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<const CBlock> pblock, bool fForceProcessing, bool *fNewBlock)
+bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<const CBlock> pblock, CValidationState& state, bool fForceProcessing, bool *fNewBlock)
```
let's rename it to `dos_state` for clarity since it seems all callers(aside from dummy usage) calls it that.

TheBlueMatt replied:
Done. Also fixed an issue where the FormatStateMessage in ActivateBestChain failure prints the state state, not the dummy_state state.

instagibbs commented on src/validation.cpp:
```diff
@@ -3449,15 +3448,14 @@ bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<cons
-    CValidationState state; // Only used to report errors, not invalidity - ignore it
-    if (!::ChainstateActive().ActivateBestChain(state, chainparams, pblock))
+    CValidationState dummy_state; // Only used to report errors, not invalidity - ignore it
+    if (!::ChainstateActive().ActivateBestChain(dummy_state, chainparams, pblock))
```
unrelated to PR but can `state` still be valid here?

instagibbs commented on src/net_processing.cpp:
```diff
@@ -1232,7 +1232,7 @@ void PeerLogicValidation::UpdatedBlockTip(const CBlockIndex *pindexNew, const CB
  * Handle invalid block rejection and consequent peer banning, maintain which
  * peers announce compact blocks.
```
Suggested enhancement to comment:

"This is called after cursory DoS checks fail, and after full validation of the block in question."

TheBlueMatt commented on src/validation.h:
```diff
@@ -198,22 +198,31 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
  *
- * May not be called in a
- * validationinterface callback.
+ * If fForceProcessing is set (or fNewBlock returns true), barring pruning and
```
But the "barring pruning..." bit applies *both* to fForceProcessing and fNewBlock being set to true upon return? I think that would change the meaning.

TheBlueMatt replied:
Hmm, maybe I misunderstood what @instagibbs was suggesting. Took the changes.

instagibbs replied:
I can't remember what my suggestion meant but the new text reads much better

TheBlueMatt commented on src/validation.h:
```diff
@@ -198,22 +198,31 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
  *
- * Note that we guarantee that either the proof-of-work is valid on pblock, or
- * (and possibly also) BlockChecked will have been called.
+ * If pblock connects, and has been mutation, state is guaranteed to be some
```
Good point, clarified it a bunch, though its pretty subtle. I hope its a bit better now.

TheBlueMatt committed:
Clarify comment in BlockChecked in net_processing a touch.

instagibbs commented:
looks like it doesn't compile anymore

TheBlueMatt commented:
Right, just an issue triggered by the rebase-on-master testing. Rebased so should work now.

jnewbery commented:
Concept ACK. Will review when the build is fixed.

TheBlueMatt commented:
Built is fixed :)

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2839,15 +2846,19 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             }
-            LOCK(cs_main); // hold cs_main for CBlockIndex::IsValid()
-            if (pindex->IsValid(BLOCK_VALID_TRANSACTIONS)) {
+            if (dos_state.IsValid()) {
```
This seems like a behaviour change. If `ConnectBlock()` fails for this block, then `pindex->nStatus` will be updated with `BLOCK_FAILED_VALID`. Previously, that would prevent this branch from being executed, but now we're only looking at the return `CValidationState` from `CheckBlock()` and `AcceptBlock()`.

If this is an intentional behaviour change, can you pull it out into a separate commit and comment on why in the commit log?

TheBlueMatt replied:
Don't think it matters either way (though also don't think it was *intentional* per se). It's not so much an important change as it is an accidental tweak thats probably a bit better but shouldn't matter. Once we get the block we have to mark it as received, but we cannot do that if the block is malleated. If the block is invalid, we don't care too much - essentially anyone who sends us the block is gonna get the banhammer (if the block is not received via compact blocks).

jnewbery replied:
If it doesn't matter either way, can you leave it unchanged? Trying to pick through what exactly `MarkBlockAsReceived()` is doing and whether changing whether it's called or not is adding quite a lot of burden to review.

TheBlueMatt replied:
IMO, this is much cleaner, even if the behavior proper doesn't matter. a) reaching into validation to figure out what happened after calling PNB is...a massive layer violation, this should be returned, not figured out by side-effects, b) it means one chunk of cs_main less (even if its only in the uncommon case), and c) it matters a lot for #16323.

jnewbery replied:
ok, if it does matter, can you separate into its own commit, with a commit log saying _why_ it matters a lot :)

TheBlueMatt replied:
Two different definitions of "matter", I suppose. Anyway, I pulled it out on its own.

jonatack replied:
Tests that clarify and assert the points in this discussion would be reassuring.

TheBlueMatt replied:
There really is no material difference in behavior here. Its much *cleaner* code, but the behavior is the same. Hence the two different definitions of "matter".

jnewbery commented on src/rpc/mining.cpp:
```diff
@@ -132,7 +132,8 @@ static UniValue generateBlocks(const CScript& coinbase_script, int nGenerate, ui
         std::shared_ptr<const CBlock> shared_pblock = std::make_shared<const CBlock>(*pblock);
-        if (!ProcessNewBlock(Params(), shared_pblock, true, nullptr))
+        CValidationState state;
+        if (!ProcessNewBlock(Params(), shared_pblock, state, true, nullptr) || !state.IsValid())
```
I don't think `|| !state.IsValid()` is necessary. `ProcessNewBlock()` cannot return true if state is invalid.

TheBlueMatt replied:
Might very well be #16323 bleeding in, but hopefully that is required soon :).

jkczyz commented on src/rpc/mining.cpp:
```diff
@@ -733,15 +734,20 @@ static UniValue submitblock(const JSONRPCRequest& request)
+        return BIP22ValidationResult(sc.state);
+    } else {
+        return BIP22ValidationResult(dos_state);
+    }
```
Consider the less verbose version:

```c++
return BIP22ValidationResult(dos_state.IsValid() ? sc.state : dos_state);
```

Or use a guard clause to be consistent with the previous `if`:

```c++
if (!dos_state.IsValid()) {
    return BIP22ValidationResult(dos_state);
}

return BIP22ValidationResult(sc.state);
```

jkczyz commented on src/validation.h:
```diff
@@ -198,22 +198,33 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
+ * May not be called in a validationinterface callback.
  *
  * @param[in]   pblock  The block we want to process.
+ * @param[out] state This may be set to an Error state if any error occurred processing them
```
Nit: Alignment

jkczyz commented on src/rpc/mining.cpp:
```diff
@@ -733,15 +734,20 @@ static UniValue submitblock(const JSONRPCRequest& request)
+        return BIP22ValidationResult(sc.state);
+    } else {
+        return BIP22ValidationResult(dos_state);
+    }
```
Actually, you can remove the duplicate check on `dos_state.IsValid()` like so:

```c++
if (!dos_state.IsValid()) {
    return BIP22ValidationResult(dos_state);
}

if (!sc.found) {
    return "inconclusive";
}

return BIP22ValidationResult(sc.state);
```

jnewbery replied:
Much better. Can you do something similar here: https://github.com/bitcoin/bitcoin/pull/16279#discussion_r333100579 please :)

TheBlueMatt replied:
Ugh, github's review tools have gone to shit. I responded at https://github.com/bitcoin/bitcoin/pull/16279#discussion_r334708611

TheBlueMatt commented on src/validation.h:
```diff
@@ -198,22 +198,33 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
+ * If the block pblock is built on is in our header tree, and pblock is a
+ * candidate for accepting to disk (either because it is a candidate for the
+ * best chain soon, or due to fForceProcessing being set), but pblock has been
+ * mutated, state is guaranteed to be some non-IsValid() state.
```
Note that it doesn't have to be a candidate for the best chain, as it can meet the "more or same work" criteria, hence the word "soon". I took the drop of "due to", but am not a huge fan of dropping the redundant subject after the parenthetical (as then you have to re-read skipping the parenthetical).

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3747,37 +3747,35 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
         if (!ret) {
-            GetMainSignals().BlockChecked(*pblock, state);
-            return error("%s: AcceptBlock FAILED (%s)", __func__, FormatStateMessage(state));
+            return error("%s: AcceptBlock FAILED (%s)", __func__, FormatStateMessage(dos_state));
```
Hmm? Note that ret may get set to false in the if (ret) {} conditional above.

fjahr replied:
I missed it too :) Maybe consider not reusing the `ret` variable but giving the return of `AcceptBlock` a different name.

jkczyz commented on src/validation.cpp:
```diff
@@ -3747,37 +3747,35 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
+            ret = ::ChainstateActive().AcceptBlock(pblock, dos_state, chainparams, &pindex, fForceProcessing, nullptr, fNewBlock);
         }
         if (!ret) {
-            GetMainSignals().BlockChecked(*pblock, state);
```
The commit message mentions this in an offhand sort of way. But it seems like this (i.e. removing the event notification) and the way callers handle the secondary return are the primary changes introduced in this commit. The commit log could better reflect this rather than describing the implementation mechanics (i.e. the 'how').

You could argue that removing the event notification should be a commit unto itself. Doing so would definitely make formulating the two commit summaries much easier.

TheBlueMatt replied:
Clarified the commit message to make it more clear. Not gonna bother splitting as it really is the same idea - we return the dos_state instead of passing it back via a callback.

jkczyz commented on src/validation.cpp:
```diff
@@ -3747,37 +3747,35 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
         if (!ret) {
-            GetMainSignals().BlockChecked(*pblock, state);
-            return error("%s: AcceptBlock FAILED (%s)", __func__, FormatStateMessage(state));
+            return error("%s: AcceptBlock FAILED (%s)", __func__, FormatStateMessage(dos_state));
```
Unrelated to this commit, the error here may be from either `CheckBlock` or `AcceptBlock`. Related to this commit, the returned `CValidationState` could also be from either rather than just `AcceptBlock` but the commit summary mentions only the latter.

I think this reflects a need for better naming / code organization rather than the need to make the commit message more verbose, however.

amitiuttarwar replied:
but since this commit is touching this code, might be worth updating error message to "AcceptBlock or CheckBlock FAILED..." to be more accurate?

TheBlueMatt replied:
Historically, CheckBlock was only called as a part of AcceptBlock, and I think we'd like to eventually go back there. See #/9765 and the bitcoin-dev mail titled "Vulnerability relating to 64-byte transactions in Bitcoin Core 0.13 branch". (sorry, lots and lots of context to this code.....)

fjahr commented on src/rpc/mining.cpp:
```diff
@@ -132,7 +132,8 @@ static UniValue generateBlocks(const CScript& coinbase_script, int nGenerate, ui
         }
         std::shared_ptr<const CBlock> shared_pblock = std::make_shared<const CBlock>(*pblock);
-        if (!ProcessNewBlock(Params(), shared_pblock, true, nullptr))
+        CValidationState state;
```
This one could also be named `dos_state`, I think?

fjahr commented:
tested ACK eabc9a192765df5b66fd4cdc65c4b645a39391cf

Reviewed code and ran tests + modified tests to try failure scenarios. While beyond the scope of this PR, something to note is that all the tests that were changed only tested the happy path. Something to improve in follow-ups, maybe after #16323.

TheBlueMatt committed:
Return the AcceptBlock CValidationState directly in ProcessNewBlock

In practice this means that CheckBlock+ContextualCheckBlock are
called with a passed-in CValidationState before we move onto
connecting the best chain. This makes conceptual sense as these
calls represent the DoS checks on a block (ie PoW and malleability)
which the caller almost certainly wants to know about right away
and shouldn't have to wait on a callback for.

Further, as other validationinterface clients shouldn't care about
someone submitting bogus malleated blocks to PNB, the BlockChecked
callback is no longer called in response to a malleated block (ie
it really does now mean "block has been checked as a part of
connecting it").

This also makes it much, much easier to move the best chain
activation logic to a background thread as it implies that if PNB
returns with a IsValid() CValidationState we don't need to care
about trying to process (non-malleated) copies of the block from
other peers.

TheBlueMatt committed:
Use dos_state to check for malleation instead of CBlockIndex state

etscrivner commented:
tested ACK 2ec121f09d8f7117fc9a8f830a7242f9a3602b78.

Compiled and ran all of the tests. Ran a testnet node with this code for ~1 day without issue. Currently running one with the updated commits from today (will post if any issues).

jonatack reviewed (commented):
Concept ACK 2ec121f09d8f7117fc9a8f830a7242f9a3602b78. Changes seem ~ok, build/tests now green, bitcoind runs without apparent issues afaict. Will need to dig deeper into the code and context to add a code review ACK.

adamjonas commented:
tested ACK 2ec121f. Compiled and ran unit and functional tests.

ariard commented on src/validation.h:
```diff
@@ -196,22 +196,33 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
+ * Performs initial sanity checks using the provided CValidationState before
+ * connecting any block(s). If you want to *possibly* get feedback on whether
+ * pblock is valid beyond just cursory mutation/DoS checks, you must install
+ * a CValidationInterface (see validationinterface.h) - this will have its
```
"The installed `BlockChecked` method will be called for *any* block completing validation. It may not be called for the provided pblock as this one not being part of the best chain. But if `BlockChecked` is called for it and its `CValidationState` is invalid, the invalidity reason will be different than the `CValidationState got back from `ProcessNewBlock`". Just a verbose attempt to underscores clearly the fact you will get 2 different set of invalidity reasons.

ariard commented on src/validation.h:
```diff
@@ -196,22 +196,33 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
+ * @param[out]  state This may be set to an Error state if any error occurred processing them
  * @param[in]   fForceProcessing Process this block even if unrequested; used for non-network block sources and whitelisted peers.
  * @param[out]  fNewBlock A boolean which is set to indicate if the block was first received via this call
  * @returns     If the block was processed, independently of block validity
```
The whole comment could be clearer by defining what we mean by processed. Like "A processed block is one saved to disk and submit as candidate to be part of the best chain." We can also add "passed the cursory mutation/Dos checks" if I'm correct will always return failed (non-processed) if block fail them. So both concepts are a bit mangled, as processed presume these preliminary checks as successful. For DoS storage risks, we don't want to process blocks without basic safeguards.

ariard commented on src/rpc/mining.cpp:
```diff
@@ -733,11 +734,15 @@ static UniValue submitblock(const JSONRPCRequest& request)
         return "duplicate";
     }
+    if (!dos_state.IsValid()) {
+        return BIP22ValidationResult(dos_state);
```
nit: slight behavior change as now we return error on weak validity check, but would say it's now more accurate as `BlockChecked` wasn't giving state of the block we wait for.

TheBlueMatt replied:
Ehh, getblocktemplate is a mess already. TBH who really cares since no one actually uses it except to get a block template and submit a block. No one uses the return value.

ariard reviewed (approved):
Code review ACK 2ec121f. Mostly suggestions to improve comments on a very mangled part of the codebase, had a look on following PRs, beyond expected performance improvements, it should be far easier to understand with a clean layer separation.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -1227,6 +1227,9 @@ void PeerLogicValidation::UpdatedBlockTip(const CBlockIndex *pindexNew, const CB
  * Handle invalid block rejection and consequent peer banning, maintain which
  * peers announce compact blocks.
+ * Called both in case of cursory DoS checks failing (implying the peer is likely
+ * sending us bogus data) and after full validation of the block fails (which may
```
The existing comment already says that.
"Handle invalid block rejection and consequent peer banning".

TheBlueMatt commented:
Oops cause I didn’t comment, this kind of trivial refactor is probably detrimental in the death-by-1000-cuts model without #16323 and #16324, which seem to be too scary to get any review, so would recommend against running with it without reviving those (which are, of course, up for grabs if someone thinks they can make progress with it).

>> On Nov 12, 2019, at 15:19, Matt Corallo <notifications@github.com> wrote:
> ﻿
> Closed #16279.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jnewbery commented:
It's very difficult for reviewers to understand the motivation and direction for this sequence of PRs, and consequently to help you move them forward toward merge. The OP for this PR states "It's just two rather-nice (IMO) cleanups.", but your final comment here is "would recommend against running with it without reviving those [other PRs]". Those comments seem to me to be in direct contraction with each other. If the changes here are rather-nice cleanups, that suggests to me that they're worth taking independently from #16323 and #16324.

TheBlueMatt commented:
Right, but nice cleanups don’t always need to be merged, and do, by themselves, carry other weight. See the regular stream of “if this gets merged, X, Y, and Z will conflict” from Drahtbot (which is awesome). If they have a goal towards enabling other work, that’s great, but by themselves, unless they make it less likely some issue crops up in the future (which I kinda doubt for these changes), they’re often not worth merging.

> On Nov 12, 2019, at 17:16, John Newbery <notifications@github.com> wrote:
> 
> ﻿
> It's very difficult for reviewers to understand the motivation and direction for this sequence of PRs, and consequently to help you move them forward toward merge. The OP for this PR states "It's just two rather-nice (IMO) cleanups.", but your final comment here is "would recommend against running with it without reviving those [other PRs]". Those comments seem to me to be in direct contraction with each other. If the changes here are rather-nice cleanups, that suggests to me that they're worth taking independently from #16323 and #16324.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ryanofsky commented:
Marking up for grabs

http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-12.html#l-455

\<jnewbery> #16279 seems to have already had a lot of review. I liked everything except the last commit in that one, so I might grab those

TheBlueMatt commented:
See above discussion. I don’t think that tag makes sense here.

> On Nov 13, 2019, at 11:57, Russell Yanofsky <notifications@github.com> wrote:
> 
> ﻿
> Marking up for grabs
> 
> http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-12.html#l-455
> 
> <jnewbery> #16279 seems to have already had a lot of review. I liked everything except the last commit in that one, so I might grab those
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jnewbery commented:
Rebased in #17479
